File: mpi/code/c/solution/trapeze.c
/*
 * calcul d'une integrale par la methode des trapezes.
 */


#include <stdlib.h>
#include <stdio.h>
#include <math.h>

double f(double x) {

  return 4.0/(1+x*x);

}

double compute_integral(double start, double dx, int nSegments)
{

  int i;
  double integral = 0.0;

  for (i=0; i<nSegments; i++) {

    double x1 = start + dx*i;
    integral += ( f(x1) + f(x1 + dx) ) / 2 * dx;

  }

  return integral;
}


int main (int argc, char* argv[])
{

  // les bornes de l'integrale
  double a,b;

  // nombre de segments total
  int n = 1024;

  // longueur d'un segment
  double dx;

  // calcul local  d'integrale
  double integral;

  a=0.0;
  b=1.0;

  // calcul les bornes d'integration locale
  dx=(b-a)/n;

  // compute  integral
  integral = compute_integral(0,dx,n);

  printf("Resultat final : %g\n",integral);
  printf("Erreur absolue : %g\n",integral-4*atan(1.0));

  return EXIT_SUCCESS;

}

File: mpi/code/fortran/solution/trapeze.f90
!
! calcul d'une integrale par la methode des trapezes.
!
! gfortran -O2 -Wall trapeze.f90 -o trapeze
!

program trapeze

  implicit none

  ! les bornes de l'integrale
  real(kind=8) :: a,b

  ! nombre de segments total
  integer :: n = 1024
 
  ! longueur d'un segment
  real(kind=8) :: dx

  ! calcul local  d'integrale
  real(kind=8) :: integral

  !a = 0.0
  !b = 1.0

  write(*,*) 'Enter a : '; read(*,*) a
  write(*,*) 'Enter b : '; read(*,*) b
  write(*,*) 'Enter n : '; read(*,*) n

  ! calcul les bornes d'integration locale
  dx = (b-a)/n

  ! compute integral
  call compute_integral(a,dx,n,integral)
  
  write(*,*) 'Resultat final : ',integral
  write(*,*) 'Erreur absolue : ',integral - (f_primitive(b) - f_primitive(a) )

contains

   function f(x) result(y)
     
     implicit none
     
     real(kind=8), intent(in) :: x
     real(kind=8)             :: y
     
     y = 4.0/(1.0+x*x)
     
   end function f

   ! utilisée a posteriori pour vérification
   function f_primitive(x) result(y)
     
     implicit none
     
     real(kind=8), intent(in) :: x
     real(kind=8)             :: y
     
     y = 4.0*atan(x)
     
   end function f_primitive

  subroutine compute_integral(start, dx, nSegments, integral)
    
    implicit none

    ! dummy variables
    real(kind=8), intent(in)  :: start
    real(kind=8), intent(in)  :: dx
    integer,      intent(in)  :: nSegments
    real(kind=8), intent(out) :: integral

    ! local variables
    integer      :: i
    real(kind=8) :: x1

    integral = 0.0

    do i=0,nSegments-1
    
       x1 = start + dx*i
       integral = integral + ( f(x1) + f(x1 + dx) ) / 2 * dx
       
    end do

  end subroutine compute_integral

end program trapeze
